File: API/PetCenterServices/Services/UserService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Storage;
using PetCenterModels.DBTables;
using PetCenterModels.DataTransferObjects;
using PetCenterModels.SearchObjects;
using PetCenterServices.Interfaces;
using PetCenterServices.Utils;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using PetCenterServices.Recommender;
using Microsoft.IdentityModel.Tokens;
using Microsoft.Extensions.Logging;
using PetCenterModels.ModelUtils;

namespace PetCenterServices.Services
{
    public class UserService : BaseCRUDService<User,UserSearchObject,UserRequestDTO,UserResponseDTO>, IUserService
    {

        private readonly IRecommenderSystem recommender;

        public UserService(PetCenterDBContext ctx,ILoggerFactory _ilogger, IRecommenderSystem rec) : base(ctx,_ilogger)
        {
            dbSet = ctx.Users;
            recommender = rec;
        }


        protected override Task<IQueryable<User>> Filter(Guid token_holder, UserSearchObject search)
        {
            IQueryable<User> output = dbSet.Include(u=>u.UserAccount).OrderBy(u=>u.Id);

            if (search.AuthoritySpecifier == Access.BusinessAccount)
            {
                output = output.Where(u=>u.UserAccount.AccessLevel==Access.BusinessAccount && u.Id!=token_holder);
            }

            if (!string.IsNullOrWhiteSpace(search.UserName))
            {
                output = output.Where(u=>u.UserName!.ToLower().StartsWith(search.UserName.ToLower()));
            }

            if (search.EmployedBy != null)
            {
                if (search.IncludeExclude)
                {
                    output = output.Where(u=> dbContext.EmployeeRecords.Any(e=>e.UserId==u.Id && e.FranchiseId==search.EmployedBy));
                }            
                else
                {
                    output = output.Where(u=> !dbContext.EmployeeRecords.Any(e=>e.UserId==u.Id && e.FranchiseId==search.EmployedBy));
                }
            }
            return Task.FromResult(output);
        }

        public override async Task<ServiceOutput<UserResponseDTO>> GetById(Guid session,Guid token_holder, Guid id, Access authorization_level, FileScope fileScope = FileScope.Invalid)
        {
            UserResponseDTO? output = UserResponseDTO.FromEntity(await dbSet.Include(u=>u.UserAccount).FirstOrDefaultAsync(u=>u.Id==id));

            if (output == null) 
            {      
                return ServiceOutput<UserResponseDTO>.Error(HttpCode.NotFound, "No user with this ID exists.");                 
            }

         
            if (authorization_level == Access.BusinessAccount)
            {
                IQueryable<Guid> records = dbContext.EmployeeRecords.Where(e=>e.UserId==token_holder).Select(e=>e.FranchiseId);
                List<Franchise> workplaces =  await dbContext.Franchises.Include(f=>f.Facilities).Include(f=>f.ShelteredAnimals).Where(f=>records.Contains(f.Id)||f.OwnerId==token_holder).OrderBy(w=>w.Id).ToListAsync();
                output.Workplaces=workplaces.Select(w=>FranchiseResponseDTO.FromEntity(w,w.OwnerId==token_holder)!).ToList();
              
                List<Guid> workplace_ids = workplaces.Select(w=>w.Id).ToList();
                output.Notifications = await dbContext.Notifications.Where(n=>n.UserId==token_holder || (n.FranchiseId!=null && workplace_ids.Contains(n.FranchiseId.Value!))).Select(n=>NotificationSubDTO.FromEntity(n)!).ToListAsync();
                
            }
            else if(authorization_level==Access.User)
            {
                List<Individual>individuals = await dbContext.IndividualAnimals.Include(i=>i.AnimalBreed).ThenInclude(b=>b.AnimalKind).Include(i=>i.MedicalRecord).Where(i=>i.Owned && i.OwnerId==token_holder).ToListAsync();
                List<MedicalProcedureSpecification> specifications = await dbContext.MedicalProcedureSpecifications.Include(m=>m.MedicalProcedure).ToListAsync();


                output.Notes=new();
                output.Notes.Add(await recommender.ShoppingList(dbContext,token_holder));
                foreach(Individual ind in individuals)
                {
                    output.Notes.AddRange(await recommender.AddNotesToPet(dbContext,ind,specifications));
                }
                output.Notifications = await dbContext.Notifications.Include(n=>n.RelevantListing).Where(n=>n.UserId==token_holder && (n.ListingId==null||(n.RelevantListing.Approved && n.RelevantListing.Visible))).Select(n=>NotificationSubDTO.FromEntity(n)!).ToListAsync();
                output.UserWishlist = await dbContext.Wishlists.Where(w=>w.UserId==output.Id).Select(w=>w.Term).ToListAsync();
                output.UserSupplies= await dbContext.SupplyRecords.Where(s=>s.UserId==token_holder).Select(s=>SuppliesSubDTO.FromEntity(s)!).ToListAsync();
                output.OwnedAnimals= individuals.Select(i=>IndividualResponseDTO.FromEntity(i)!).ToList();
            }           
            
            return ServiceOutput<UserResponseDTO>.Success(output);
        }

        public async Task<ServiceOutput<Guid>> GetUserState(Guid token_holder)
        {
            User? usr = await dbSet.FindAsync(token_holder);
            if(usr==null){return ServiceOutput<Guid>.Error(HttpCode.NotFound,"No user with this ID exists.");}
            return ServiceOutput<Guid>.Success(usr.UserState);
        }

        public async Task<ServiceOutput<List<AnnouncementSubDTO>>> GetAnnouncements(Access role)
        {
            IQueryable<Announcement> query = dbContext.Announcements;

            if (role == Access.User)
            {
                query=query.Where(a=>a.UserVisible);
            }
            else if (role == Access.BusinessAccount)
            {
                query=query.Where(a=>a.BusinessVisible);
            }

            List<Announcement> list = await query.ToListAsync();

            return ServiceOutput<List<AnnouncementSubDTO>>.Success(list.Select(e=> AnnouncementSubDTO.FromEntity(e)!).ToList());
            
        }

        public async Task<ServiceOutput<List<ReportResponseSubDTO>>> GetReports()
        {
            
            List<Report> list = await dbContext.Reports.ToListAsync();

            return ServiceOutput<List<ReportResponseSubDTO>>.Success(list.Select(e=> ReportResponseSubDTO.FromEntity(e)!).ToList());
            
        }


        public override async Task<ServiceOutput<UserResponseDTO>> Put(Guid session,Guid token_holder,UserRequestDTO ent)
        {

            User? current = await dbContext.Users.FindAsync(ent.Id);

            if (current != null)
            {
            await using(IDbContextTransaction tx = await dbContext.Database.BeginTransactionAsync())
                {
                    try
                    {
                        current.CurrentVersion=ent.CurrentVersion;
                        current.UserName = ent.UserName;
                        await dbContext.SaveChangesAsync();
                        await tx.CommitAsync();

                        return ServiceOutput<UserResponseDTO>.Success(UserResponseDTO.FromEntity(current));
                        
                    }
                    catch(Exception ex)
                    {
                        await tx.RollbackAsync();
                        return ServiceOutput<UserResponseDTO>.FromException(ex,logger);
                    }
                }

              

            }

            return ServiceOutput<UserResponseDTO>.Error(HttpCode.NotFound,"No user with this ID exists.");

        }

        public override Task<ServiceOutput<UserResponseDTO>> Post(Guid session,Guid token_holder,UserRequestDTO ent)
        {
            return Task.FromResult(ServiceOutput<UserResponseDTO>.Error(HttpCode.NotImplemented,"Illegal endpoint."));
        }

        public async Task<ServiceOutput<string>> SetEmployee(Guid caller_id, Guid usr_id, Guid franchise_id, bool add_remove)
        {

            await using(IDbContextTransaction tx = await dbContext.Database.BeginTransactionAsync(IsolationLevel.Serializable))
            {

                try
                {
                    User? usr =  await dbContext.Users.Include(u=>u.UserAccount).FirstOrDefaultAsync(u=>u.Id==usr_id);
                    User? owner = await dbContext.Users.FindAsync(caller_id);
                    Franchise? franchise = await dbContext.Franchises.FindAsync(franchise_id);
                    EmployeeRecord? record = await dbContext.EmployeeRecords.FirstOrDefaultAsync(r=>r.UserId==usr_id && r.FranchiseId==franchise_id);

                    if(usr==null || owner==null || franchise==null || usr.UserAccount==null)
                    {
                        
                        return ServiceOutput<string>.Error(HttpCode.NotFound,"One or more resources needed for this operation are missing.");
                    }

                    if (add_remove)
                    {
                        if (franchise.OwnerId != owner.Id)
                        {
                            
                            return ServiceOutput<string>.Error(HttpCode.Forbidden,"The token holder does not own the specified franchise.");
                        }

                        if (usr.UserAccount.AccessLevel != Access.BusinessAccount)
                        {
                            
                            return ServiceOutput<string>.Error(HttpCode.BadRequest,"The specified user is not eligible to be an employee.");
                        }

                        if(caller_id == usr_id)
                        {
                         
                            return ServiceOutput<string>.Error(HttpCode.BadRequest,"The owner of a franchise is already considered an employee.");
                        }

                        if (record == null)
                        {
                            EmployeeRecord newRecord = new EmployeeRecord
                            {
                                UserId = usr_id,
                                FranchiseId = franchise_id,                       
                            };

                            usr.UserState=Guid.NewGuid();

                            await dbContext.EmployeeRecords.AddAsync(newRecord);
                            await dbContext.SaveChangesAsync();
                            await tx.CommitAsync();
                        }

                        return ServiceOutput<string>.Success("Employee hired successfully.");
                    }
                    else
                    {

                        if (franchise.OwnerId == usr_id)
                        {
                           
                            return ServiceOutput<string>.Error(HttpCode.BadRequest,"The owner of a franchise cannot remove themselves from the employee list.");
                        }

                        if (franchise.OwnerId != owner.Id && owner.Id!=usr_id)
                        {
                           
                            return ServiceOutput<string>.Error(HttpCode.Forbidden,"You are not allowed to perform this operation.");
                        }

                        if(record!=null)
                        {
                            dbContext.EmployeeRecords.Remove(record);     

                            usr.UserState=Guid.NewGuid();
                       
                            await dbContext.SaveChangesAsync();
                            await tx.CommitAsync();
                        }

                        return ServiceOutput<string>.Success("Employee removed successfully.");
                    }
                   
                            
                }
                catch(Exception ex)
                {
                    
                    await tx.RollbackAsync();
                    return ServiceOutput<string>.FromException(ex,logger);
                }


            }


        }

        public override async Task<ServiceOutput<object>> IsClearedToUpdate(Guid token_holder, UserRequestDTO resource)
        {
            if (!resource.Validate())
            {
                return ServiceOutput<object>.Error(HttpCode.BadRequest,"Request validation failure.");
            }

            if (resource.Id != token_holder)
            {
                return ServiceOutput<object>.Error(HttpCode.Forbidden,"Token does not belong to user.");
            }

            User? existing = await dbSet.FirstOrDefaultAsync(u=>u.UserName==resource.UserName && u.Id!=resource.Id);

            if (existing != null)
            {
                return ServiceOutput<object>.Error(HttpCode.Conflict,$"The username {resource.UserName} is already taken.");
            }

            return ServiceOutput<object>.Success(null,HttpCode.NoContent);
        }

        public override Task<ServiceOutput<object>> IsClearedToDelete(Guid token_holder, Guid resourceId)
        {                   
            if(resourceId!=token_holder)
            {
                return Task.FromResult(ServiceOutput<object>.Error(HttpCode.Forbidden,"Token does not belong to user."));
            }

            return Task.FromResult(ServiceOutput<object>.Success(null,HttpCode.NoContent));
        }



        public async Task<ServiceOutput<string>> SetWishlistTerm(Guid usr_id, string term, bool add_remove)
        {
            term = term.ToLowerInvariant();
            Wishlist? existing = await dbContext.Wishlists.FirstOrDefaultAsync(w=>w.UserId==usr_id && w.Term==term);


            if (add_remove)
            {
                if(existing == null)
                {
                    Wishlist newEntry = new Wishlist
                    {
                        UserId = usr_id,
                        Term = term
                    };

                    await dbContext.Wishlists.AddAsync(newEntry);
                    await dbContext.SaveChangesAsync();
                }
                return ServiceOutput<string>.Success("Term added to wishlist.");
            }
            else
            {
                if (existing != null)
                {
                await using(IDbContextTransaction tx = await dbContext.Database.BeginTransactionAsync())
                    {
                        try{
                            await existing.StageDeletion<Wishlist>(dbContext,dbContext.Wishlists);
                            await dbContext.SaveChangesAsync();
                            await tx.CommitAsync();
                        }
                        catch(Exception ex){await tx.RollbackAsync(); return ServiceOutput<string>.FromException(ex,logger);}
                    }
                }
                return ServiceOutput<string>.Success("Term removed from wishlist.");
            }

            
        }


        public async Task<ServiceOutput<AnnouncementSubDTO>> AddAnnouncement(string body, bool user_visible, bool business_visible, int expiry)
        {
            if (string.IsNullOrWhiteSpace(body))
            {
                return ServiceOutput<AnnouncementSubDTO>.Error(HttpCode.BadRequest,"Announcement body cannot be empty.");
            }

            Announcement? existing = await dbContext.Announcements.FirstOrDefaultAsync(a=>a.Body.ToLower()==body.ToLower() && a.UserVisible==user_visible && a.BusinessVisible==business_visible);
            if (existing != null)
            {
                existing.Expiry = DateTime.UtcNow.AddDays(expiry);             
                await dbContext.SaveChangesAsync();
                StaticDataVersionHolder.AnnouncementVersion=Guid.NewGuid();
                return ServiceOutput<AnnouncementSubDTO>.Success(AnnouncementSubDTO.FromEntity(existing));
            }
           
            Announcement newAnnouncement = new Announcement
            {
                Body = body,
                UserVisible = user_visible,
                BusinessVisible = business_visible,
                Expiry = DateTime.UtcNow.AddDays(expiry)
            };

            try
            {
                await dbContext.Announcements.AddAsync(newAnnouncement);
                await dbContext.SaveChangesAsync();
            }
            catch(Exception ex)
            {
                return ServiceOutput<AnnouncementSubDTO>.FromException(ex,logger);
            }

            StaticDataVersionHolder.AnnouncementVersion=Guid.NewGuid();
            return ServiceOutput<AnnouncementSubDTO>.Success(AnnouncementSubDTO.FromEntity(newAnnouncement),HttpCode.Created);
        }


        public async Task<ServiceOutput<string>> RemoveAnnouncement(Guid announcement_id)
        {
            Announcement? existing = await dbContext.Announcements.FindAsync(announcement_id);

            if (existing != null)
            {
            await using(IDbContextTransaction tx = await dbContext.Database.BeginTransactionAsync())
                {

                    try
                    {
                        await existing.StageDeletion<Announcement>(dbContext,dbContext.Announcements);
                        await dbContext.SaveChangesAsync();
                        StaticDataVersionHolder.AnnouncementVersion=Guid.NewGuid();
                        await tx.CommitAsync();
                    }
                    catch(Exception ex)
                    {
                        await tx.RollbackAsync();
                        return ServiceOutput<string>.FromException(ex,logger);
                    }

                }
                
            }

            return ServiceOutput<string>.Success("Announcement removed successfully.");
        }

        public async Task<ServiceOutput<NotificationSubDTO>> AddNotification(Guid token_holder,Access auth, string title,string body,Guid userId,Guid? franchiseId,Guid? listingId,int expiry)
        {
            if (string.IsNullOrWhiteSpace(title) || string.IsNullOrWhiteSpace(body))
            {
                return ServiceOutput<NotificationSubDTO>.Error(HttpCode.BadRequest,"Notification title and body cannot be empty.");
            }

            User? usr = await dbSet.FindAsync(userId);
            if (usr == null)
            {
                return ServiceOutput<NotificationSubDTO>.Error(HttpCode.NotFound,"The specified user does not exist.");
            }
            

            Franchise? franch = null;

            if (franchiseId != null)
            {
                franch = await dbContext.Franchises.FindAsync(franchiseId);

                if (franch == null)
                {
                    return ServiceOutput<NotificationSubDTO>.Error(HttpCode.NotFound,"Could not find specified franchise.");
                }
            }


            if(auth==Access.BusinessAccount && (franch == null||franch.OwnerId!=token_holder))
            {
                return ServiceOutput<NotificationSubDTO>.Error(HttpCode.NotFound,"You do not own a franchise with the provided ID.");
            }
            


            await using (IDbContextTransaction tx = await dbContext.Database.BeginTransactionAsync()){

                try
                {
                    Notification notif = new();

                    Notification? existing = await dbContext.Notifications
                        .FirstOrDefaultAsync(n =>
                            n.UserId == userId &&
                            n.ListingId == listingId &&
                            n.FranchiseId == franchiseId &&
                            n.Title.ToLower() == title.ToLower() &&
                            n.Body.ToLower() == body.ToLower()
                        );

                    if (existing != null)
                    {
                        existing.Expiry = DateTime.UtcNow.AddDays(expiry);

                        notif = existing;
                        
                    }
                    else{

                        Notification newNotification = new()
                        {
                            Title = title,
                            Body = body,
                            UserId = userId,
                            ListingId = listingId,
                            FranchiseId = franchiseId,
                            Expiry = DateTime.UtcNow.AddDays(expiry)
                        };

                        await dbContext.Notifications.AddAsync(newNotification);

                        notif= newNotification;
                    }

                  

                    if (franch != null)
                    {
                        
                            
                        List<User> emp = await dbContext.Users.Where(u=> (dbContext.Franchises.Any(f=>f.Id==franchiseId && f.OwnerId==u.Id)||dbContext.EmployeeRecords.Any(e=>e.UserId==u.Id&&e.FranchiseId==franchiseId))).ToListAsync();

                        foreach(User e in emp)
                        {
                            e.UserState=Guid.NewGuid();
                        }

                    }

                    await dbContext.SaveChangesAsync();
                    await tx.CommitAsync();

                    HttpCode code = (existing!=null)? HttpCode.OK:HttpCode.Created;

                    return ServiceOutput<NotificationSubDTO>.Success(NotificationSubDTO.FromEntity(notif),code);
                }
                catch (Exception ex)
                {
                    await tx.RollbackAsync();
                    return ServiceOutput<NotificationSubDTO>.FromException(ex,logger);
                }

            }
        }

        public async Task<ServiceOutput<string>> RemoveNotification(Guid notification_id)
        {
            Notification? existing = await dbContext.Notifications.FindAsync(notification_id);

            if (existing != null)
            {

            await using(IDbContextTransaction tx = await dbContext.Database.BeginTransactionAsync())
                {
                    try
                    {
                        await existing.StageDeletion<Notification>(dbContext,dbContext.Notifications);
                        await dbContext.SaveChangesAsync();
                        await tx.CommitAsync();
                    }
                    catch(Exception ex)
                    {
                        await tx.RollbackAsync();
                        return ServiceOutput<string>.FromException(ex,logger);
                    }
                }
                
            }

            return ServiceOutput<string>.Success("Notification removed successfully.");
        }
       
    }
}

File: Frontend/pet_center_app/lib/services/user_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:pet_center_app/models/data_transfer/listing/sub_dtos.dart';
import 'package:pet_center_app/models/data_transfer/user/user_request_dto.dart';
import 'package:pet_center_app/models/data_transfer/user/user_response_dto.dart';
import 'package:pet_center_app/services/static_user_data_service.dart';

import 'package:pet_center_app/utils/app_config.dart';
import 'package:pet_center_app/utils/app_style.dart';
import 'package:pet_center_app/utils/globals.dart';
import 'package:pet_center_app/utils/jwt_parser.dart';
import 'package:pet_center_app/utils/service_output.dart';

class UserService {
  static Future<UserResponseDTO?> getSelf() async {
    apiServiceBusy.value = true;
    try {
      final response = await http.get(
        Uri.parse("${AppConfig.apiBaseUrl}/api/User/Me"),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<UserResponseDTO>(
        response,
        (json) => UserResponseDTO.fromJson(json as Map<String, dynamic>),
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<String?> setEmployee(
    String userId,
    String franchiseId,
    bool hire,
  ) async {
    apiServiceBusy.value = true;
    try {
      final query = <String, String>{};
      query['add_remove'] = hire.toString();

      final response = await http.put(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/SetEmployee/$userId/$franchiseId",
        ).replace(queryParameters: query),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<String>(
        response,
        (json) => (json as Map<String, dynamic>)['value'] as String,
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<String?> getUserStatus() async {
    apiServiceBusy.value = true;
    try {
      final response = await http.get(
        Uri.parse("${AppConfig.apiBaseUrl}/api/User/Status"),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<String>(
        response,
        (json) => (json as Map<String, dynamic>)['value'] as String,
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<UserResponseDTO?> update(UserRequestDTO input) async {
    apiServiceBusy.value = true;
    try {
      final response = await http.put(
        Uri.parse("${AppConfig.apiBaseUrl}/api/User/${self?.id}"),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: jsonEncode(input.toJson()),
      );

      final result = await ServiceOutput.fromResponse<UserResponseDTO>(
        response,
        (json) => UserResponseDTO.fromJson(json as Map<String, dynamic>),
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<List<AnnouncementSubDTO>?> getAnnouncements() async {
    apiServiceBusy.value = true;
    try {
      final response = await http.get(
        Uri.parse("${AppConfig.apiBaseUrl}/api/User/Announcement"),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<List<AnnouncementSubDTO>>(
        response,
        (json) => (json as List)
            .map((e) => AnnouncementSubDTO.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<int?> countReports() async {
    apiServiceBusy.value = true;
    try {
      final response = await http.get(
        Uri.parse("${AppConfig.apiBaseUrl}/api/User/Report/Count"),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<int>(
        response,
        (json) => (json as Map<String, dynamic>)['value'] as int,
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<List<ReportResponseSubDTO>?> getReports(int page) async {
    apiServiceBusy.value = true;
    try {
      final query = <String, String>{};
      query['page'] = page.toString();

      final response = await http.get(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/Report",
        ).replace(queryParameters: query),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result =
          await ServiceOutput.fromResponse<List<ReportResponseSubDTO>>(
            response,
            (json) => (json as List)
                .map(
                  (e) =>
                      ReportResponseSubDTO.fromJson(e as Map<String, dynamic>),
                )
                .toList(),
          );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<List<ReportResponseSubDTO>?> getAllReports() async {
    final pageCount = await countReports();

    if (pageCount == null) {
      return null;
    }

    List<ReportResponseSubDTO> output = [];
    final seen = <String?>{};

    for (int i = 0; i < pageCount; i++) {
      final newEntries = await getReports(i);

      if (newEntries == null) {
        return null;
      }

      for (final ent in newEntries) {
        if (seen.add(ent.id)) {
          output.add(ent);
        }
      }
    }

    return output;
  }

  static Future<int?> count(
    bool includeExclude,
    String? userName,
    String? employedBy,
  ) async {
    apiServiceBusy.value = true;
    try {
      final query = <String, String>{};
      query['page'] = 0.toString();

      query['includeExclude'] = includeExclude.toString();
      if (employedBy != null) {
        query['employedBy'] = employedBy;
      }
      if (userName != null && userName.trim() != "") {
        query['userName'] = userName;
      }

      final response = await http.get(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/Count",
        ).replace(queryParameters: query),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<int>(
        response,
        (json) => (json as Map<String, dynamic>)['value'] as int,
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<List<UserResponseDTO>?> get(
    bool includeExclude,
    String? userName,
    String? employedBy,
    int page,
  ) async {
    apiServiceBusy.value = true;
    try {
      final query = <String, String>{};
      query['page'] = page.toString();
      if (userName != null && userName.trim() != "") {
        query['userName'] = userName;
      }
      query['includeExclude'] = includeExclude.toString();
      if (employedBy != null) {
        query['employedBy'] = employedBy;
      }

      final response = await http.get(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User",
        ).replace(queryParameters: query),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<List<UserResponseDTO>>(
        response,
        (json) => (json as List)
            .map((e) => UserResponseDTO.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<bool> reset() async {
    apiServiceBusy.value = true;
    try {
      final response = await http.delete(
        Uri.parse("${AppConfig.apiBaseUrl}/api/User/${self?.id}"),
        headers: {'Authorization': 'Bearer $rawToken'},
      );

      apiServiceBusy.value = false;
      return ServiceOutput.isSuccess(response);
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return false;
    }
  }

  static Future<String?> setWishlistTerm(String term) async {
    apiServiceBusy.value = true;
    try {
      final query = <String, String>{'add_remove': true.toString()};

      final response = await http.put(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/SetTerm/$term",
        ).replace(queryParameters: query),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<String>(
        response,
        (json) => (json as Map<String, dynamic>)['value'] as String,
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<bool> removeWishlistTerm(String term) async {
    apiServiceBusy.value = true;
    try {
      final query = <String, String>{'add_remove': false.toString()};

      final response = await http.put(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/SetTerm/$term",
        ).replace(queryParameters: query),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      apiServiceBusy.value = false;
      return ServiceOutput.isSuccess(response);
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return false;
    }
  }

  static Future<AnnouncementSubDTO?> addAnnouncement(
    String announcement,
    bool userVisible,
    bool businessVisible,
    int daysValid,
  ) async {
    apiServiceBusy.value = true;
    try {
      final query = <String, String>{
        'user_visible': userVisible.toString(),
        'business_visible': businessVisible.toString(),
        'days_valid': daysValid.toString(),
      };

      final response = await http.put(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/Announcement",
        ).replace(queryParameters: query),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: jsonEncode({'text': announcement}),
      );

      final result = await ServiceOutput.fromResponse<AnnouncementSubDTO>(
        response,
        (json) => AnnouncementSubDTO.fromJson(json as Map<String, dynamic>),
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<String?> removeAnnouncement(String announcementId) async {
    apiServiceBusy.value = true;
    try {
      final response = await http.delete(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/Announcement/$announcementId",
        ),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<String>(
        response,
        (json) => (json as Map<String, dynamic>)['value'] as String,
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<NotificationSubDTO?> addNotification(
    String userId,
    String title,
    String body,
    String? franchiseId,
    String? listingId,
    int daysValid,
  ) async {
    apiServiceBusy.value = true;
    try {
      final query = <String, String>{'days_valid': daysValid.toString()};
      if (franchiseId != null) query['franchise_id'] = franchiseId;
      if (listingId != null) query['listing_id'] = listingId;

      final response = await http.put(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/Notification/$userId",
        ).replace(queryParameters: query),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({'title': title, 'body': body}),
      );

      final result = await ServiceOutput.fromResponse<NotificationSubDTO>(
        response,
        (json) => NotificationSubDTO.fromJson(json as Map<String, dynamic>),
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<String?> removeNotification(String notificationId) async {
    apiServiceBusy.value = true;
    try {
      final response = await http.delete(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/Notification/$notificationId",
        ),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<String>(
        response,
        (json) => (json as Map<String, dynamic>)['value'] as String,
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }

  static Future<bool?> setSeen(String notifId) async {
    apiServiceBusy.value = true;
    try {
      final response = await http.put(
        Uri.parse(
          "${AppConfig.apiBaseUrl}/api/User/Notification/Seen/$notifId",
        ),
        headers: {
          'Authorization': 'Bearer $rawToken',
          'Accept': 'application/json',
        },
      );

      final result = await ServiceOutput.fromResponse<bool>(
        response,
        (json) => (json as Map<String, dynamic>)['value'] as bool,
      );

      apiServiceBusy.value = false;
      return result;
    } catch (ex) {
      showError(ex);
      apiServiceBusy.value = false;
      return null;
    }
  }
}
